core/kernel/interrupt.c

```diff
@@ -23,27 +23,21 @@ void itr_init(struct itr_chip *chip)
 	itr_chip = chip;
 }
 
-static struct itr_handler *find_handler(size_t it)
-{
-	struct itr_handler *h;
-
-	SLIST_FOREACH(h, &handlers, link)
-		if (h->it == it)
-			return h;
-	return NULL;
-}
-
 void itr_handle(size_t it)
 {
-	struct itr_handler *h = find_handler(it);
-
-	if (!h) {
-		EMSG("Disabling unhandled interrupt %zu", it);
-		itr_chip->ops->disable(itr_chip, it);
-		return;
+	struct itr_handler *h = NULL;
+	bool was_handled = false;
+
+	SLIST_FOREACH(h, &handlers, link) {
+		if (h->it == it) {
+			if (h->handler(h) == ITRR_HANDLED)
+				was_handled = true;
+			else if (!(h->flags & ITRF_SHARED))
+				break;
+		}
 	}
 
-	if (h->handler(h) != ITRR_HANDLED) {
+	if (!was_handled) {
 		EMSG("Disabling interrupt %zu not handled by handler", it);
```

Comment:
suggest to update the error message: `EMSG("Disabling unhandled interrupt %zu", it);`